File: tests/ui/crud/read-shares.js
const {chromium} = require('playwright')
const path = require('path')
const {
  login,
  getConfig,
  createTestWallet,
  ensureWalletExists
} = require('../auth-helper')

/**
 * Test: Read wallet shares
 * This test retrieves and displays existing wallet shares
 */

;(async () => {
  const browser = await chromium.launch({headless: true, slowMo: 500})
  const page = await browser.newPage()
  const config = getConfig()

  let success = false

  try {
    console.log('🚀 Starting read wallet shares test...')

    // Step 1: Login as admin
    console.log('📝 Step 1: Logging in as admin...')
    await login(page)

    // Step 2: Ensure user has a wallet
    console.log('📝 Step 2: Ensuring user has a wallet...')
    await page.goto(`${config.baseUrl}/wallet`)
    await page.waitForTimeout(2000)

    const existingWallet = await ensureWalletExists(page)
    const existingAdminKey = existingWallet.adminkey

    // Step 3: Create a fresh test wallet via API
    console.log('📝 Step 3: Creating fresh test wallet via API...')
    const testWallet = await createTestWallet(page, existingAdminKey)
    if (!testWallet) {
      throw new Error('Failed to create test wallet')
    }

    const walletId = testWallet.id
    const adminKey = testWallet.adminkey

    console.log(`✅ Using test wallet ID: ${walletId}`)

    // Step 4: Read shares via API
    console.log('📝 Step 4: Reading shares via API...')
    if (adminKey) {
      try {
        const response = await page.request.get(
          `${config.baseUrl}/api/v1/wallet_shares/${walletId}`,
          {
            headers: {'X-Api-Key': adminKey}
          }
        )

        if (response.ok()) {
          const shares = await response.json()
          const count = Array.isArray(shares) ? shares.length : 0

          console.log(`📊 Found ${count} share(s)`)

          if (Array.isArray(shares) && shares.length > 0) {
            shares.forEach((share, index) => {
              console.log(`\n📄 Share ${index + 1}:`)
              console.log(`   ID: ${share.id}`)
              console.log(`   User: ${share.username || share.user_id}`)
              console.log(`   Permissions: ${share.permissions}`)
              console.log(`   Shared by: ${share.shared_by}`)
              console.log(`   Shared at: ${share.shared_at}`)
              console.log(`   Accepted: ${share.accepted}`)
            })
          }

          success = true
          console.log('\n✅ SUCCESS! Shares read successfully')
        } else {
          console.log(`❌ API returned ${response.status()}`)
        }
      } catch (error) {
        console.log('❌ Error reading shares:', error.message)
      }
    } else {
      console.log('❌ No admin API key available')
    }

    // Step 4: Verify UI shows shares
    console.log('\n📝 Step 4: Checking UI for Share Wallet button...')
    // The Share Wallet button is a round button with group icon in top right
    const shareButton = page.locator('button.text-deep-purple i.material-icons:has-text("group")').locator('..')

    if (await shareButton.isVisible({timeout: 5000})) {
      await shareButton.click()
      await page.waitForTimeout(2000)

      // Take screenshot of shares dialog
      await page.screenshot({
        path: path.join(__dirname, '../screenshots/read-shares-dialog.png'),
        fullPage: true
      })
      console.log('📸 Screenshot saved: read-shares-dialog.png')

      // Check for "Current Shares" section
      const currentSharesSection = page.locator(
        'text=Current Shares, text=Shared Users'
      )
      if (await currentSharesSection.isVisible()) {
        console.log('✅ Current Shares section visible in UI')
      }

      console.log('✅ UI verification complete')
    }
  } catch (error) {
    console.error('💥 Error:', error.message)
    await page.screenshot({
      path: path.join(__dirname, '../screenshots/read-shares-error.png'),
      fullPage: true
    })
  } finally {
    await browser.close()
  }

  process.exit(success ? 0 : 1)
})()

File: tests/api/read-shares.py
#!/usr/bin/env python3
"""
Test: Read wallet shares via API
This test retrieves and displays wallet shares using the REST API
"""

import asyncio
import os
import httpx
from loguru import logger

# Load config from .env.local
def load_config():
    config = {}
    env_path = os.path.join(os.path.dirname(__file__), '../../.env.local')

    if os.path.exists(env_path):
        with open(env_path, 'r') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#') and '=' in line:
                    key, value = line.split('=', 1)
                    config[key.strip()] = value.strip()

    return {
        'base_url': config.get('TEST_LNBITS_URL', os.getenv('TEST_LNBITS_URL')),
        'admin_key': config.get('TEST_ADMIN_API_KEY', os.getenv('TEST_ADMIN_API_KEY')),
        'wallet_id': config.get('TEST_WALLET_ID', os.getenv('TEST_WALLET_ID'))
    }

async def test_read_shares():
    """Test reading wallet shares"""
    config = load_config()

    if not config['admin_key']:
        logger.error("❌ TEST_ADMIN_API_KEY must be set in .env.local")
        return False

    if not config['wallet_id']:
        logger.error("❌ TEST_WALLET_ID must be set in .env.local")
        return False

    logger.info("🚀 Starting read wallet shares API test...")
    logger.info(f"📍 Base URL: {config['base_url']}")
    logger.info(f"📍 Wallet ID: {config['wallet_id']}")

    async with httpx.AsyncClient() as client:
        # Get wallet shares
        logger.info("📝 Fetching wallet shares...")
        response = await client.get(
            f"{config['base_url']}/api/v1/wallet_shares/{config['wallet_id']}",
            headers={'X-Api-Key': config['admin_key']}
        )

        if response.status_code != 200:
            logger.error(f"❌ Failed to get shares: {response.status_code}")
            logger.error(f"   Response: {response.text}")
            return False

        shares = response.json()

        if not isinstance(shares, list):
            logger.error(f"❌ Unexpected response format: {type(shares)}")
            return False

        logger.info(f"📊 Found {len(shares)} share(s)")

        if len(shares) > 0:
            logger.info("\n📄 Share Details:")
            for i, share in enumerate(shares, 1):
                logger.info(f"\n  Share {i}:")
                logger.info(f"    ID: {share['id']}")
                logger.info(f"    User: {share.get('username', share['user_id'])}")
                logger.info(f"    Permissions: {share['permissions']}")
                logger.info(f"    Shared by: {share['shared_by']}")
                logger.info(f"    Shared at: {share['shared_at']}")
                logger.info(f"    Accepted: {share['accepted']}")

                # Decode permissions
                perms = []
                if share['permissions'] & 1: perms.append('VIEW')
                if share['permissions'] & 2: perms.append('CREATE_INVOICE')
                if share['permissions'] & 4: perms.append('PAY_INVOICE')
                if share['permissions'] & 8: perms.append('MANAGE_SHARES')
                logger.info(f"    Permission flags: {', '.join(perms)}")
        else:
            logger.info("ℹ️  No shares found for this wallet")

        logger.info("\n🎉 SUCCESS! Shares read successfully!")
        return True

if __name__ == '__main__':
    success = asyncio.run(test_read_shares())
    exit(0 if success else 1)
